sections/design/basics-of-react.tex

```diff
@@ -0,0 +1,97 @@
+This variable declaration can be called a React element which can then be rendered to the DOM
+Other variables can be declared and used inside JSX by wrapping it in brackets.
+\begin{center}
+    \texttt{ReactDOM.render(element, document.getElementById('root'));}
+\end{center}
+We can render the element to the DOM by using \texttt{ReactDom.render()}. The element is passed to this function as well as the DOM node we would like it to be rendered to. In this case we are rendering the react element \texttt{element} to the DOM node \texttt{root}.
+\begin{lstlisting}
+const name = 'Josh Perez';
+const element = <h1>Hello, {name}</h1>;
+
+ReactDOM.render(
+  element,
+  document.getElementById('root')
+);
+\end{lstlisting}
+The variable that is embedded in the JSX can contain any valid JavaScript expression.
+\\\\
+In React we use components, which is a single file that holds all of the logic, html and styling that we need.
+
+\begin{lstlisting}
+import React from "react";
+    
+function App() {
+    return (
+        <div>
+            <h1>Hello React</h1>
+        </div>
+    );
+}
+
+export default App
+\end{lstlisting}
+In this example we create the component \texttt{App} and export it. By exporting it we can use the \texttt{App} component elsewhere and render it. 
+In the function we return JSX which React will then compile to html.
+The idea is to build the UI from components by exporting them and then importing them into other larger components to contain them. This means that the components act as the building blocks of the UI because we can reuse them in other components.
+\begin{lstlisting}
+import React from "react";
+        
+function Button() {
+    return (
+        <div>
+            <button onClick={console.log('Hello')}>Hello React</button>
+        </div>
+    );
+}
+    
+export default Button
+\end{lstlisting}
+As an example we can create a button component with an on-click event embedded into it. 
+Here we use brackets to enable us to embed JavaScript in the JSX which will be run when the button is clicked.
+We can now export this button and it can be imported by other components. 
+A component is conceptually a JavaScript function, which means that they can accept inputs called \texttt{props} which can contain data that can be used in the react elements. If a component is used in another component, we can use \texttt{props} to pass data to the child component.
+\\\\
+In general, \texttt{props} should be read-only, which means a component should try to change it's values. 
+To allow the UI elements to change over time we instead make use of \textit{State} which allows the components to change their output over time when handling user inputs or network responses. 
+% Skriv om stateful og stateless components
+\textit{State}  is similar to \texttt{props} but is instead private and only the specific component can make changes to it. To make use of \textit{State} we need to initialize the component as a class instead of a function
```

Comment:
```suggestion
\textit{State}  is similar to \texttt{props} but is instead private and only the specific component can make changes to it. To make use of \textit{State} we need to initialize the component as a class instead of a function
```
```suggestion
\textit{State} is similar to \texttt{props} but is instead private and only the specific component can make changes to it.
To make use of \textit{State} we need to initialize the component as a class instead of a function.
```